File: spring-boot-webflux-api-rest/src/main/java/com/springboot/webflux/restapi/app/models/documents/Product.java
package com.springboot.webflux.restapi.app.models.documents;

import java.util.Date;

import javax.validation.Valid;
import javax.validation.constraints.NotEmpty;
import javax.validation.constraints.NotNull;

import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.format.annotation.DateTimeFormat;

//sudo service mongod start
/**
 * 
 * @author Mario Ruiz Rojo
 * It represents products collection
 */
@Document(collection="products")
public class Product {
	/**
	 * Identifier of product
	 */
	@Id
	private String id;
	
	/**
	 * Name of product
	 */
	@NotEmpty
	private String name;
	
	/**
	 * Price of product
	 */
	@NotNull
	private Double price;
	
	/**
	 * Creation date of product
	 */
	@DateTimeFormat(pattern="yyyy-MM-dd")
	private Date createdAt;
	
	/**
	 * Category of product
	 */
	@Valid
	@NotNull
	private Category category;
	
	/**
	 * Picture file name of product
	 */
	private String picture;
	
	public Product() {
		super();
	}
	
	public Product(String name, Double price) {
		super();
		this.name = name;
		this.price = price;
	}
	
	public Product(String name, Double price, Category category) {
		this(name,price);
		this.category = category;
	}
	
	public Product(Product productReceived) {
		this(productReceived.name,productReceived.price,productReceived.category);
		this.picture=productReceived.picture;
	}

	/**
	 * 
	 * @return Identifier of product
	 */
	public String getId() {
		return id;
	}
	
	/**
	 * 
	 * @return id to set Identifier of product
	 */
	public void setId(String id) {
		this.id = id;
	}
	
	/**
	 * 
	 * @return Name of product
	 */
	public String getName() {
		return name;
	}
	/**
	 * 
	 * @param name to set Name of product
	 */
	public void setName(String name) {
		this.name = name;
	}
	
	/**
	 * 
	 * @return Price of product
	 */
	public Double getPrice() {
		return price;
	}
	/**
	 * 
	 * @param price to set Price of product
	 */
	public void setPrice(Double price) {
		this.price = price;
	}
	/**
	 * 
	 * @return Creation date of product
	 */
	public Date getCreatedAt() {
		return createdAt;
	}
	/**
	 * 
	 * @param createdAt set to Creation date of product
	 */
	public void setCreatedAt(Date createdAt) {
		this.createdAt = createdAt;
	}

	/**
	 * 
	 * @return Category of product
	 */
	public Category getCategory() {
		return category;
	}

	/**
	 * 
	 * @param category set to Category of product
	 */
	public void setCategory(Category category) {
		this.category = category;
	}

	/**
	 * 
	 * @return Picture file name of product
	 */
	public String getPicture() {
		return picture;
	}

	/**
	 * 
	 * @param picture to Picture file name of product
	 */
	public void setPicture(String picture) {
		this.picture = picture;
	}
}

File: crud-react-webpage/src/models/Product.ts
import {Category} from './Category'

/**
 * It represents product data
 */
export class Product{
    /**
     * Identifier of product 
     */
    private id: string;   
    /**
     * Name of product 
     */ 
    private name: string; 
    /**
     * Price of product 
     */    
    private price: number;  
    /**
     * Creation date of product 
     */   
    private createdAt: Date;   
    /**
     * Category of product 
     */  
    private category: Category;  
    /**
     * Picture file name of product 
     */   
    private picture: string;    

    /**
     * Constructor from jsonObject
     * @param product jsonObject
     */
    constructor(product : any){
        if(product==undefined){
            this.id = '';
            this.name = '';
            this.price = 0.0;
            this.createdAt = new Date();
            this.category = new Category(undefined);
            this.picture = '';
        }else{
            this.id = product.id;
            this.name = product.name;
            try{
                this.price = parseFloat(product.price);
            }
            catch(e){
                this.price = 0.0
            }
            this.createdAt = new Date(product.createdAt);
            this.category = new Category(product.category);
            this.picture = product.picture;
        }        
    }

    /**
     * 
     * @returns Identifier
     */
    public getId(): string {
        return this.id;
    }
    /**
     * 
     * @param value to set Identifier of product
     */
    public setId(value: string) {
        this.id = value;
    }
    /**
     * 
     * @returns Name of product
     */
    public getName(): string {
        return this.name;
    }
    /**
     * 
     * @param value to set Name of product
     */
    public setName(value: string) {
        this.name = value;
    }
    /**
     * 
     * @returns Price of product
     */
    public getPrice(): number {
        return this.price;
    }
    /**
     * 
     * @param value to set Price of product
     */
    public setPrice(value: number) {
        this.price = value;
    }
    /**
     * 
     * @returns Creation date of product
     */
    public getCreatedAt(): Date {
        return this.createdAt;
    }
    /**
     * 
     * @param value to set Creation date of product
     */
    public setCreatedAt(value: Date) {
        this.createdAt = value;
    }
    /**
     * 
     * @returns Category
     */
    public getCategory(): Category {
        return this.category;
    }
    /**
     * 
     * @param value to set Category of product
     */
    public setCategory(value: Category) {
        this.category = value;
    }
    /**
     * 
     * @returns Picture file name of product
     */
    public getPicture(): string {
        return this.picture;
    }
    /**
     * 
     * @param value to set Picture file name of product
     */
    public setPicture(value: string) {
        this.picture = value;
    }
}